File: leetcode/march2024/python/16_contiguous_array.py
# INTUITION
# Kadane's algorithm variation

# APPROACH
# Check out Kadane's algorithm

# TIME COMPLEXITY
# TC: O(n)

# SPACE COMPLEXITY
# SC: O(n) - for hashmap

# CODE

class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        index = {}
        index[0] = -1
        len = cnt = 0
        for i, val in enumerate(nums):
            cnt += 1 if val == 1 else -1
            if cnt in index:
                len = max(len, i - index[cnt])
            else:
                index[cnt] = i
        return len

File: leetcode/march2024/c++/16_contiguous_array.cpp
// INTUITION
// Kadane's algorithm variation

// APPROACH
// Check out Kadane's algorithm

// TIME COMPLEXITY
// TC: O(n)

// SPACE COMPLEXITY
// SC: O(n) - for hashmap

// CODE

class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, int> index;
        index[0] = -1;
        int cnt = 0, len = 0;
        for(int i = 0; i < n; i++) {
            cnt = cnt + (nums[i] == 1 ? 1 : -1);
            if(index.find(cnt) != index.end()) {
                len = max(len, i - index[cnt]);
            } else {
                index[cnt] = i;
            }
        }
        return len;
    }
};